File: checkdigit/dispatcher.py
"""
dispatcher.py
=============
Detects the format of an uploaded payload from its CONTENT (not its file
extension) and routes it to the matching corrector. Detection is ordered and
fail-loud: an XML document that is not recognized container XML, and any input
we cannot confidently place, is reported rather than silently mishandled. Plain
text is the catch-all -- anything with no structured markers is scanned for bare
container numbers.

Returns (Detection, CorrectionReport). For unsupported input the report is None
and Detection.reason explains why; the service records that as a rejected event.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional, Set, Tuple

import edifact_corrector
import snx_corrector
import x12_corrector
import txt_corrector
from correction_report import CorrectionReport


@dataclass
class Detection:
    fmt: str            # "edifact" | "snx" | "x12" | "txt" | "unsupported"
    reason: str = ""


def detect_format(text: str) -> Detection:
    s = text.lstrip("\ufeff \t\r\n")
    head = s[:1000]

    # 1) XML -> terminal container XML ("SNX" family), matched structurally
    #    (namespace/prefix-agnostic: <snx>, <ns:snx>, discharge-list, container/unit)
    if s.startswith("<?xml") or s.startswith("<"):
        if (re.search(r"<(?:[A-Za-z][\w.-]*:)?snx[\s/>]", head)
                or "<line-discharge-list" in head
                or ("<container" in s and "eqid=" in s)
                or ("<unit" in s and "unique-key=" in s)):
            return Detection("snx")
        return Detection("unsupported", "XML detected but not a recognized container XML schema")

    # 2) EDIFACT -- service-string-advice or interchange/message header
    if s.startswith("UNA") or s.startswith("UNB") or "UNB+" in head or "UNH+" in head:
        return Detection("edifact")

    # 3) X12 -- interchange/group/transaction header with segment terminator
    if "~" in s and (s.startswith("ISA") or "ISA*" in head or "GS*" in head or "ST*" in head):
        return Detection("x12")

    # 4) plain text catch-all
    return Detection("txt")


def correct(text: str, *, owner_policy: str = "strict", trust: bool = False,
            known_owner_prefixes: Optional[Set[str]] = None, policy=None
            ) -> Tuple[Detection, Optional[CorrectionReport]]:
    det = detect_format(text)
    if det.fmt == "edifact":
        return det, edifact_corrector.correct_edifact(text, owner_policy=owner_policy, policy=policy)
    if det.fmt == "snx":
        return det, snx_corrector.correct_snx(text, owner_policy=owner_policy, policy=policy)
    if det.fmt == "x12":
        return det, x12_corrector.correct_x12(text, owner_policy=owner_policy, policy=policy)
    if det.fmt == "txt":
        return det, txt_corrector.correct_txt(
            text, trust=trust, owner_policy=owner_policy,
            known_owner_prefixes=known_owner_prefixes, policy=policy)
    return det, None        # unsupported


# ── opt-in structured-text formats (CSV/TSV, fixed-width) ─────────────────────
# These are NEVER auto-detected: a CSV of container numbers is valid plain text,
# and silently treating a column as trusted would violate the low-trust default.
# The caller opts in with a format hint + the column spec; absent that, such a
# file flows through the safe free-text path. parse_options carries the spec.
def correct_with_hint(text: str, *, format_hint: str, parse_options: Optional[dict] = None,
                      owner_policy: str = "strict", trust: bool = True,
                      known_owner_prefixes: Optional[Set[str]] = None, policy=None
                      ) -> Tuple[Detection, Optional[CorrectionReport]]:
    """Route an explicitly-hinted structured-text format. Supported hints:
       'csv'  : parse_options = {columns:[...], delimiter?, has_header?, whole_cell?}
       'fixed': parse_options = {ranges:[[start,end],...], header_lines?, whole_field?}
    Unknown hints fall back to content-based detection (correct())."""
    import csv_corrector
    import fixedwidth_corrector
    from csv_locator import CsvError
    from fixedwidth_locator import FixedWidthError
    opts = dict(parse_options or {})

    if format_hint == "csv":
        cols = opts.get("columns")
        if not cols:
            return Detection("unsupported",
                             "CSV hint requires parse_options.columns (indices or header names)"), None
        try:
            rep = csv_corrector.correct_csv(
                text, cols, trust=trust, owner_policy=owner_policy,
                delimiter=opts.get("delimiter"), has_header=opts.get("has_header", True),
                whole_cell=opts.get("whole_cell", True),
                known_owner_prefixes=known_owner_prefixes, policy=policy)
        except CsvError as exc:
            return Detection("unsupported", f"CSV column spec error: {exc}"), None
        return Detection("csv"), rep

    if format_hint in ("fixed", "fixedwidth", "fixed_width"):
        ranges = opts.get("ranges")
        if not ranges:
            return Detection("unsupported",
                             "fixed-width hint requires parse_options.ranges [[start,end],...]"), None
        try:
            rep = fixedwidth_corrector.correct_fixedwidth(
                text, [tuple(r) for r in ranges], trust=trust, owner_policy=owner_policy,
                header_lines=opts.get("header_lines", 0),
                whole_field=opts.get("whole_field", True),
                known_owner_prefixes=known_owner_prefixes, policy=policy)
        except FixedWidthError as exc:
            return Detection("unsupported", f"fixed-width spec error: {exc}"), None
        return Detection("fixed"), rep

    # unknown hint -> safe content-based path
    return correct(text, owner_policy=owner_policy, trust=trust,
                   known_owner_prefixes=known_owner_prefixes, policy=policy)


# ── binary front door ─────────────────────────────────────────────────────────
_ZIP_MAGIC = b"PK\x03\x04"


def detect_bytes(data: bytes) -> "Detection | None":
    """
    Sniff binary containers BEFORE any text decode (a ZIP 'decodes' fine under
    latin-1 and would otherwise be garbage-scanned as txt). Returns a Detection
    for recognized binary formats ('xlsx') or explicit unsupported binaries;
    returns None when the payload should continue down the text path.
    """
    if data[:4] == _ZIP_MAGIC:
        import io
        import zipfile
        try:
            with zipfile.ZipFile(io.BytesIO(data)) as zf:
                names = set(zf.namelist())
        except zipfile.BadZipFile:
            return Detection(fmt="unsupported", reason="corrupt ZIP archive")
        if "xl/workbook.xml" in names:
            return Detection(fmt="xlsx")
        if "word/document.xml" in names:
            return Detection(fmt="unsupported",
                             reason="Word document (.docx) is not a supported format")
        if "ppt/presentation.xml" in names:
            return Detection(fmt="unsupported",
                             reason="PowerPoint (.pptx) is not a supported format")
        return Detection(fmt="unsupported",
                         reason="ZIP archive that is not an Excel workbook")
    if data[:5] == b"%PDF-":
        return Detection(fmt="unsupported", reason="PDF is not a supported format")
    return None

File: src/lib/dispatcher.ts
import type { PolicyLike } from "./checkdigit";
import type { CorrectionReport } from "./report";
import { correctCsv, CsvError, type ColumnSel } from "./formats/csv";
import { correctEdifact } from "./formats/edifact";
import { correctFixedWidth, FixedWidthError, type ColRange } from "./formats/fixedwidth";
import { correctSnx } from "./formats/snx";
import { correctTxt } from "./formats/txt";
import { correctX12 } from "./formats/x12";

/*
 * Port of checkdigit/dispatcher.py: content-based format detection, the binary
 * front door, and routing to the format correctors. Plain text is the
 * catch-all; CSV and fixed-width are never auto-detected (the caller opts in
 * with a hint and a column spec).
 */

export type Format = "edifact" | "snx" | "x12" | "txt" | "csv" | "fixed" | "xlsx" | "unsupported";

export interface Detection {
  fmt: Format;
  reason: string;
}

// dispatcher.py:33-56
export function detectFormat(text: string): Detection {
  const s = text.replace(/^[﻿ \t\r\n]+/, "");
  const head = s.slice(0, 1000);
  if (s.startsWith("<?xml") || s.startsWith("<")) {
    if (
      /<(?:[A-Za-z][\w.-]*:)?snx[\s/>]/.test(head) ||
      head.includes("<line-discharge-list") ||
      (s.includes("<container") && s.includes("eqid=")) ||
      (s.includes("<unit") && s.includes("unique-key="))
    ) {
      return { fmt: "snx", reason: "" };
    }
    return { fmt: "unsupported", reason: "XML detected but not a recognized container XML schema" };
  }
  if (s.startsWith("UNA") || s.startsWith("UNB") || head.includes("UNB+") || head.includes("UNH+")) {
    return { fmt: "edifact", reason: "" };
  }
  if (s.includes("~") && (s.startsWith("ISA") || head.includes("ISA*") || head.includes("GS*") || head.includes("ST*"))) {
    return { fmt: "x12", reason: "" };
  }
  return { fmt: "txt", reason: "" };
}

const ZIP_MAGIC = [0x50, 0x4b, 0x03, 0x04];

/** Member names from a ZIP central directory (no decompression). Throws on a corrupt archive. */
export function zipMemberNames(data: Uint8Array): string[] {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const minEocd = 22;
  if (data.length < minEocd) throw new Error("archive too small");
  let eocd = -1;
  for (let i = data.length - minEocd; i >= Math.max(0, data.length - minEocd - 65535); i--) {
    if (view.getUint32(i, true) === 0x06054b50) {
      eocd = i;
      break;
    }
  }
  if (eocd === -1) throw new Error("end of central directory not found");
  const entries = view.getUint16(eocd + 10, true);
  const cdSize = view.getUint32(eocd + 12, true);
  const cdOffset = view.getUint32(eocd + 16, true);
  if (cdOffset + cdSize > data.length) throw new Error("central directory outside the file");
  const names: string[] = [];
  let p = cdOffset;
  const decoder = new TextDecoder("utf-8");
  for (let k = 0; k < entries; k++) {
    if (p + 46 > data.length || view.getUint32(p, true) !== 0x02014b50) throw new Error("bad central directory entry");
    const nameLen = view.getUint16(p + 28, true);
    const extraLen = view.getUint16(p + 30, true);
    const commentLen = view.getUint16(p + 32, true);
    names.push(decoder.decode(data.subarray(p + 46, p + 46 + nameLen)));
    p += 46 + nameLen + extraLen + commentLen;
  }
  return names;
}

// dispatcher.py:134-161
export function detectBytes(data: Uint8Array): Detection | null {
  if (data.length >= 4 && ZIP_MAGIC.every((b, i) => data[i] === b)) {
    let names: string[];
    try {
      names = zipMemberNames(data);
    } catch {
      return { fmt: "unsupported", reason: "corrupt ZIP archive" };
    }
    if (names.includes("xl/workbook.xml")) return { fmt: "xlsx", reason: "" };
    if (names.includes("word/document.xml")) return { fmt: "unsupported", reason: "Word document (.docx) is not a supported format" };
    if (names.includes("ppt/presentation.xml")) return { fmt: "unsupported", reason: "PowerPoint (.pptx) is not a supported format" };
    return { fmt: "unsupported", reason: "ZIP archive that is not an Excel workbook" };
  }
  if (data.length >= 5 && data[0] === 0x25 && data[1] === 0x50 && data[2] === 0x44 && data[3] === 0x46 && data[4] === 0x2d) {
    return { fmt: "unsupported", reason: "PDF is not a supported format" };
  }
  return null;
}

export interface CorrectOptions {
  owner_policy?: string;
  trust?: boolean;
  known_owner_prefixes?: ReadonlySet<string> | null;
  policy?: PolicyLike | null;
}

// dispatcher.py:59-73
export function correct(text: string, options: CorrectOptions = {}): [Detection, CorrectionReport | null] {
  const det = detectFormat(text);
  const ownerPolicy = options.owner_policy ?? "strict";
  const policy = options.policy ?? null;
  if (det.fmt === "edifact") return [det, correctEdifact(text, { owner_policy: ownerPolicy, policy })];
  if (det.fmt === "snx") return [det, correctSnx(text, { owner_policy: ownerPolicy, policy })];
  if (det.fmt === "x12") return [det, correctX12(text, { owner_policy: ownerPolicy, policy })];
  if (det.fmt === "txt") {
    return [
      det,
      correctTxt(text, {
        trust: options.trust ?? false,
        owner_policy: ownerPolicy,
        known_owner_prefixes: options.known_owner_prefixes ?? null,
        policy,
      }),
    ];
  }
  return [det, null];
}

export interface ParseOptions {
  columns?: ColumnSel[];
  delimiter?: string | null;
  has_header?: boolean;
  whole_cell?: boolean;
  ranges?: ColRange[];
  header_lines?: number;
  whole_field?: boolean;
}

// dispatcher.py:81-127
export function correctWithHint(
  text: string,
  formatHint: string,
  parseOptions: ParseOptions = {},
  options: CorrectOptions = {},
): [Detection, CorrectionReport | null] {
  const ownerPolicy = options.owner_policy ?? "strict";
  const trust = options.trust ?? true;
  const shared = { trust, owner_policy: ownerPolicy, known_owner_prefixes: options.known_owner_prefixes ?? null, policy: options.policy ?? null };
  if (formatHint === "csv") {
    const cols = parseOptions.columns;
    if (!cols || cols.length === 0) {
      return [{ fmt: "unsupported", reason: "CSV hint requires parse_options.columns (indices or header names)" }, null];
    }
    try {
      const rep = correctCsv(text, cols, {
        ...shared,
        delimiter: parseOptions.delimiter ?? null,
        has_header: parseOptions.has_header ?? true,
        whole_cell: parseOptions.whole_cell ?? true,
      });
      return [{ fmt: "csv", reason: "" }, rep];
    } catch (err) {
      if (err instanceof CsvError) return [{ fmt: "unsupported", reason: `CSV column spec error: ${err.message}` }, null];
      throw err;
    }
  }
  if (formatHint === "fixed" || formatHint === "fixedwidth" || formatHint === "fixed_width") {
    const ranges = parseOptions.ranges;
    if (!ranges || ranges.length === 0) {
      return [{ fmt: "unsupported", reason: "fixed-width hint requires parse_options.ranges [[start,end],...]" }, null];
    }
    try {
      const rep = correctFixedWidth(text, ranges, {
        ...shared,
        header_lines: parseOptions.header_lines ?? 0,
        whole_field: parseOptions.whole_field ?? true,
      });
      return [{ fmt: "fixed", reason: "" }, rep];
    } catch (err) {
      if (err instanceof FixedWidthError) return [{ fmt: "unsupported", reason: `fixed-width spec error: ${err.message}` }, null];
      throw err;
    }
  }
  return correct(text, { ...options, trust });
}
